src/protocols/bdx/BdxTransferServer.cpp

```diff
@@ -0,0 +1,207 @@
+            ReturnLogErrorOnFailure(Responder::PrepareForTransfer(mSystemLayer, kBdxRole, flags, kMaxBdxBlockSize, kBdxTimeout));
+        }
+    }
+
+    return TransferFacilitator::OnMessageReceived(ec, payloadHeader, std::move(payload));
+}
+
+CHIP_ERROR BDXTransferServer::OnMessageToSend(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(mExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    auto & msgTypeData  = event.msgTypeData;
+    bool isStatusReport = msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport);
+
+    // All messages sent from the Sender expect a response, except for a StatusReport which would indicate an error and
+    // the end of the transfer.
+    Messaging::SendFlags sendFlags;
+    VerifyOrDo(isStatusReport, sendFlags.Set(Messaging::SendMessageFlags::kExpectResponse));
+
+    // If there's an error sending the message, close the exchange by calling Reset.
+    auto err = mExchangeCtx->SendMessage(msgTypeData.ProtocolId, msgTypeData.MessageType, std::move(event.MsgData), sendFlags);
+    VerifyOrDo(CHIP_NO_ERROR == err, Reset());
+
+    return err;
+}
+
+CHIP_ERROR BDXTransferServer::OnTransferSessionBegin(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+    VerifyOrReturnError(nullptr != mDelegate, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(mTransferProxy.Init(&mTransfer));
+    return mDelegate->OnTransferBegin(&mTransferProxy);
+}
+
+CHIP_ERROR BDXTransferServer::OnTransferSessionEnd(CHIP_ERROR error)
+{
+    assertChipStackLockedByCurrentThread();
+    VerifyOrReturnError(nullptr != mDelegate, CHIP_ERROR_INCORRECT_STATE);
+
+    LogErrorOnFailure(mDelegate->OnTransferEnd(&mTransferProxy, error));
+    Reset();
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BDXTransferServer::OnBlockReceived(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+    VerifyOrReturnError(nullptr != mDelegate, CHIP_ERROR_INCORRECT_STATE);
+
+    ByteSpan blockData(event.blockdata.Data, event.blockdata.Length);
+    return mDelegate->OnTransferData(&mTransferProxy, blockData);
+}
+
+void BDXTransferServer::AbortTransferOnFailure(CHIP_ERROR error)
+{
+    VerifyOrReturn(CHIP_NO_ERROR != error);
+    LogErrorOnFailure(error);
+    LogErrorOnFailure(mTransfer.AbortTransfer(GetBdxStatusCodeFromChipError(error)));
+}
+
+void BDXTransferServer::Reset()
+{
+    assertChipStackLockedByCurrentThread();
+
+    Responder::ResetTransfer();
+
+    if (mExchangeCtx)
+    {
+        mIsExchangeClosing = true;
+        mExchangeCtx->Close();
+        mIsExchangeClosing = false;
+        mExchangeCtx       = nullptr;
+    }
+
+    mTransferProxy.Reset();
+}
+
+void BDXTransferServer::OnExchangeClosing(Messaging::ExchangeContext * ec)
+{
+    VerifyOrReturn(!mIsExchangeClosing);
+    mExchangeCtx = nullptr;
+    LogErrorOnFailure(OnTransferSessionEnd(CHIP_ERROR_INTERNAL));
```

Comment:
This needs some documentation explaining why we know this is an error condition (presumably because mIsExchangeClosing is false so we know this is not _us_ resetting things but something else broke)?

Reply:
Done.